lib/rbnacl/hmac/sha512.rb

```diff
@@ -22,23 +22,86 @@ class SHA512 < Auth
       sodium_constant :BYTES
       sodium_constant :KEYBYTES
 
-      sodium_function :auth_hmacsha512,
-                      :crypto_auth_hmacsha512,
-                      %i[pointer pointer ulong_long pointer]
+      sodium_function :auth_hmacsha512_init,
+                      :crypto_auth_hmacsha512_init,
+                      %i[pointer pointer size_t]
 
-      sodium_function :auth_hmacsha512_verify,
-                      :crypto_auth_hmacsha512_verify,
-                      %i[pointer pointer ulong_long pointer]
+      sodium_function :auth_hmacsha512_update,
+                      :crypto_auth_hmacsha512_update,
+                      %i[pointer pointer ulong_long]
+
+      sodium_function :auth_hmacsha512_final,
+                      :crypto_auth_hmacsha512_final,
+                      %i[pointer pointer]
+
+      # Create instance without checking key length
+      #
+      # RFC 2104 HMAC
+      # The key for HMAC can be of any length.
+      #
+      # see https://tools.ietf.org/html/rfc2104#section-3
+      def initialize(key)
+        @key = Util.check_hmac_key(key, "#{self.class} key")
+        @state = State.new
+        @authenticator = Util.zeros(tag_bytes)
+
+        self.class.auth_hmacsha512_init(@state, key, key.bytesize)
+      end
+
+      # Compute authenticator for message
+      #
+      # @params [#to_str] message message to construct an authenticator for
+      def update(message)
+        self.class.auth_hmacsha512_update(@state, message, message.bytesize)
+        self.class.auth_hmacsha512_final(@state.clone, @authenticator)
```

Comment:
Ditto here re: https://github.com/crypto-rb/rbnacl/pull/166/files#r167456561

Reply:
here is the answer https://github.com/crypto-rb/rbnacl/pull/166#discussion_r167588963